_posts/2021-06-26-complete-guide-to-cors.md

```diff
@@ -0,0 +1,441 @@
+- No event listeners are registered on any [XMLHttpRequestUpload](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/upload) object
+- No ReadableStream object is used in the request
+
+The browser sends the simple request as a normal request similar to the Same Origin request, and the `Access-Control-Allow-Origin` header is checked by the browser when the response is returned.
+
+
+### Preflight Requests
+In contrast to simple requests, the browser sends preflight requests for operations that intend to change the state of existing resources in the server. We use the HTTP methods `PUT` and `DELETE` for these operations. 
+
+These requests are not considered safe so the web browser first makes sure that cross-origin communication is allowed by first sending a preflight request before sending the actual request. Requests which do not satisfy the criteria for simple request also fall under this category.
+
+The preflight request is an HTTP `OPTIONS` method which is sent automatically by the browser to the server hosting the cross-origin resource, to check that the server will permit the actual request. Along with the preflight request, the browser sends the following headers:
+
+**Access-Control-Request-Method**: This is a list of HTTP methods of the request (e.g., `GET`, `POST`, `PUT`, `DELETE`)
+**Access-Control-Request-Headers**: This is a list of headers that will be sent with the request
+**Origin**: The origin header that contains the source origin of the request
+
+The actual request to the server will not be sent if the result of the `OPTIONS` method is that the request cannot be made.
+
+After the preflight request is complete, the actual PUT method with CORS headers is sent.
+
+### CORS Requests with Credentials
+
+In most real-life situations, we need to send CORS requests loaded with some kind of access credentials which could be an `Authorization` header or cookies. The default behavior of cross-origin resource requests is for the requests to be passed without any of these credentials. 
+
+If credentials are passed with the request, the browser will not allow access to the response unless the server sends a CORS header `Access-Control-Allow-Credentials` with a value of `true`.
+
+## Example of Working with CORS
+For observing the CORS requests, let us run two web applications written in Node.Js which will communicate with each other by following the CORS standard:
+1. A `OrderProcessor` web application containing a REST API with `GET` and `PUT` methods
+2. A web application hosting the HTML page which will communicate with the REST APIs in the `OrderProcessor` application.
+
+We can run these applications in our local machine using `npm` and `node`. The application hosting the HTML page is running on `http://localhost:9000`. This makes Ajax calls with the `XMLHttpRequest` to the `OrderProcessor` application running on `http://localhost:8000`. 
```

Comment:
Again, please make clear which application is the Origin server and which is the cross-origin server.